### ISSUE TITLE
Docker MC server commands

### ISSUE THREAD

**Issue by lubocode** [gekigek99/minecraft-vanilla-server-hibernation#16]
Change from screen to systemctl for Docker version to follow master.
Further testing required before changes can be implemented.

**Comment by gekigek99** [comment]
I've read fastly your branch and I have a doubt...
Shouldn't the python/golang program be outside the docker container so that it can freeze and unfreeze the container containing the minecraft server?

**Comment by gekigek99** [comment]
An other problem that I thought... The mine server must be started (30sec delay for example) the first time a player connects.

Then when there are no players the container freezes but what happens if someone shutdown the physical machine while the container is frozen? The world won't be saved right? (A possible solution would be that everytime before the "freeze container" action a "save" command is issued but that only solves the saves problem and not the clean shutdown of the Minecraft server)

**Comment by lubocode** [comment]
Yes, the Docker image itself cannot hibernate without manual action at the moment.
However, not everyone will want to have a script running directly on the host to freeze the container.
As such, I have opted to first create an image, that starts and stops the MC server, like the original script does on a host system.

The go version already includes the functionality, that when the Docker container sends a termination signal, the script runs the function to shutdown the MC server via the defer command.
As for the python script, one would have to implement continuous listening for signals in the script and then run the shutdown function as a direct call.

Coming back to the Docker container, I do currently not know what the default behaviour of the Docker service is, when there are frozen containers and the Docker service receives a termination signal. That is, if the host system even sends such a signal instead of just simply shutting down (most probably do send a signal).

As said above, freezing and unfreezing would require another script, that runs directly on the host and executes commands over the command line of the system to interface with the Docker service.
This script would also need to differentiate between Windows and the various Linux shells to choose the correct command to run. If that is too complicated, using the Docker API would probably make more sense. 
Also, the user would probably need to specify the container ID for the script to freeze the correct container, when there are multiple instances of the Docker image installed and running as containers.

How exactly the external script would work together with the one in the image or if it would be wiser to use that external script with a "dumb" MC server Docker is another question for later.

**Comment by gekigek99** [comment]
> As such, I have opted to first create an image, that starts and stops the MC server, like the original script does on a host system.

yep sure it seems reasonable to make progress step by step

> The go version already includes the functionality, that when the Docker container sends a termination signal, the script runs the function to shutdown the MC server via the defer command.

cool! good job with that!

> freezing and unfreezing would require another script, that runs directly on the host and executes commands over the command line of the system to interface with the Docker service.

i was actually thinking of using a docker container that just contains the minecraft server and then (on the host itself) execute the minecraft server hibernation script: instead of "minecraft-server start" one would execute "docker unfreeze minecraft-container"
without the need for 2 scripts (one inside and one outside the container)

> the user would probably need to specify the container ID

I thought that during container setup script was possible to rename the container to a user specified name

**Comment by lubocode** [comment]
> > the user would probably need to specify the container ID
> 
> I thought that during container setup script was possible to rename the container to a user specified name
Yes, that is indeed possible, and when there are no duplicate names, on can address them with that (I think) but one would still have to specify the name in the script.

That script though is something I do not have on my priority list.
At the moment, I will maybe do some tinkering to optimize the docker and fix problems, if some arise.

>i was actually thinking of using a docker container that just contains the minecraft server and then (on the host itself) execute the minecraft server hibernation script: instead of "minecraft-server start" one would execute "docker unfreeze minecraft-container"
>without the need for 2 scripts (one inside and one outside the container)
Yes, I understood that. Since I run my server on unRAID it is something that was not feasible for my application and as such, I went the other route.
Now, having such a script for handling the docker would be more useful for really slow machines. I mean, what are a few seconds of waiting time until the MC server starts up?
Having options is always good though and if I suddenly find myself with a lot of time, I may look into it, if it is not done yet :)

**Comment by gekigek99** [comment]
> what are a few seconds of waiting time until the MC server starts up?
yes sure it is just a sort of cool feature just to emulate a little bit the minecraft realm.

So the python and go script remain almost identical to the "main" branch but they are wrapped up in a docker image, right?

**Comment by lubocode** [comment]
> So the python and go script remain almost identical to the "main" branch but they are wrapped up in a docker image, right?

Not quite. My first try was creating a docker with the python version, but I quickly noticed that copying files to the volume during the build process is not a viable solution. For that one would need to specify another script as the starting point which then would check if the python script is already on the volume and then start it.
So I looked into argument parsing, since the script would not be easily accessible for the user inside the docker and also reset with every update of the image. This seemed good, until I noticed that the MC server would have to be shut down somehow, when the docker is shut down.
Here I had a first look at the go script and quickly found the solution with the defer statement.

TL;DR:
Python script is basically unchanged and there is no working python docker.
Go script has added argument parsing for the MC server .jar and shuts down the MC server, when the script is terminated.



Of course, the master has since then been updated to use systemctl instead of screen, so that is not updated in either script of the docker branch. Which this issue is originally for^^

The python docker can go two routes from here:
Either create another script, which will run upon docker start and make sure the python script is on the volume that is linked with the docker container. That script would then also run the python script on startup. Here the user modifies the script itself for the parameters of the mc server.
Or the python script is left inside the docker container and arguments are passed to the script the same way as for the go docker. Here, the python script would need to be extended with argument parsing, which is not hard to do.
In BOTH cases the python script needs to be extended with a subroutine that constantly listens for a termination signal, for when the docker container shuts down, and then shuts down the mc server, since there is no comparable functionality like defer from go in python.

**Comment by lubocode** [comment]
@gekigek99 as for using systemctl instead of screen, how do you pass the arguments to the mc server .jar?
And is `systemctl start minecraft-server` a predefined command, or does one need to manually create it?

**Comment by gekigek99** [comment]
> This seemed good, until I noticed that the MC server would have to be shut down somehow, when the docker is shut down.

should't this be solvable by putting the server.jar file in the docker? -->(i mean: when you shut down the docker you shut down everything)

> The python docker can go two routes from here:

In my opinion if you make them as similar as possible it's easyer to debug, understand and use for users (and for us ;) )

> In BOTH cases the python script needs to be extended with a subroutine that constantly listens for a termination signal, for when the docker container shuts down, and then shuts down the mc server, since there is no comparable functionality like defer from go in python.

yea sure its a sort of python implementation of the defer function

> for using systemctl instead of screen, how do you pass the arguments to the mc server .jar?

on my system I have made a service for both minecraft server and hibernation script just to make it cleaner. (the service for minecraft server launches a .sh file that execute the server with all the parameters I want)

> And is systemctl start minecraft-server a predefined command, or does one need to manually create it?

I had to create a service called minecraft-server which is quite straight forward but probably for the average user is just more convenient to use screen (since it doesn't involve the management of other scripts)

**Comment by lubocode** [comment]
>should't this be solvable by putting the server.jar file in the docker? -->(i mean: when you shut down the docker you shut down everything)

Yeah, but then you would inhibit the possibility of the user using their own .jar for modding purposes. So that is not an option I would consider.

>In my opinion if you make them as similar as possible it's easyer to debug, understand and use for users (and for us ;) )

Sure. Sounds sensible.

>on my system I have made a service for both minecraft server and hibernation script just to make it cleaner. (the service for minecraft server launches a .sh file that execute the server with all the parameters I want)

Ah. Ok, then I won't be using that in the Docker.
This also means that this issue is redundant then^^
I will close it for now.
